demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,487 @@
+                          pPublishInfo->topicNameLength,
+                          callbackRecordList[ listIndex ].pTopicFilter,
+                          callbackRecordList[ listIndex ].topicFilterLength ) == true ) )
+        {
+            LogInfo( ( "Invoking subscription callback of matching topic filter: "
+                       "TopicFilter=%.*s, TopicName=%.*s",
+                       callbackRecordList[ listIndex ].topicFilterLength,
+                       callbackRecordList[ listIndex ].pTopicFilter,
+                       pPublishInfo->topicNameLength,
+                       pPublishInfo->pTopicName ) );
+
+            /* Invoke the callback associated with the record as the topics match. */
+            callbackRecordList[ listIndex ].callback( pContext, pPublishInfo );
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+SubscriptionManagerStatus_t SubscriptionManager_RegisterCallback( const char * pTopicFilter,
+                                                                  uint16_t topicFilterLength,
+                                                                  SubscriptionManagerCallback_t callback )
+{
+    assert( pTopicFilter != NULL );
+    assert( topicFilterLength != 0 );
+    assert( callback != NULL );
+
+    SubscriptionManagerStatus_t returnStatus = SUBSCRIPTION_MANAGER_INVALID;
+    size_t availableIndex = MAX_SUBSCRIPTION_CALLBACK_RECORDS;
+    bool recordExists = false;
+    size_t index = 0u;
+
+    /* Search for the first available spot in the list to store the record, and also check if
+     * a record for the topic filter already exists. */
+    while( ( recordExists == false ) && ( index < MAX_SUBSCRIPTION_CALLBACK_RECORDS ) )
+    {
+        /* Check if the index represents an empty spot in the registry. If we had already
+         * found an empty spot in the list, we will not update it. */
+        if( ( availableIndex == MAX_SUBSCRIPTION_CALLBACK_RECORDS ) &&
+            ( callbackRecordList[ index ].pTopicFilter == NULL ) )
+        {
+            availableIndex = index;
+        }
+
+        /* Check if the current record's topic filter in the registry matches the topic filter
+         * we are trying to register. */
+        else if( ( callbackRecordList[ index ].topicFilterLength == topicFilterLength ) &&
+                 ( strncmp( pTopicFilter, callbackRecordList[ index ].pTopicFilter, topicFilterLength )
+                   == 0 ) )
+        {
+            recordExists = true;
+        }
+
+        index++;
+    }
+
+    if( recordExists == true )
+    {
+        /* The record for the topic filter already exists. */
+        LogError( ( "Failed to register callback: Record for topic filter already exists: TopicFilter=%.*s",
+                    topicFilterLength,
+                    pTopicFilter ) );
+
+        returnStatus = SUBSCRIPTION_MANAGER_RECORD_EXISTS;
+    }
+    else if( ( availableIndex == MAX_SUBSCRIPTION_CALLBACK_RECORDS ) )
```

Comment:
I think the second set of parenthesis are extraneous
```suggestion
    else if( availableIndex == MAX_SUBSCRIPTION_CALLBACK_RECORDS )
```